### ISSUE TITLE
Support the seperated control of the replicating and the collecting of gradients of DataParallel to accelerate it

### ISSUE THREAD

**Issue by anoidgit** [pytorch/pytorch#13824]
## 🚀 Feature: Support seperated control of the replicating and the collecting of gradients of DataParallel to accelerate it

Add two function to `DataParallel`, one to replicate models to GPUs, the other to collect gradients.
## Motivation
I update the parameters of the designed model after several iterative forward and backward calls, which means that the `DataParallel` does not need to replicate `self.module` in every forward call, and it only needs to accumulate the gradients on the other GPUs when I am going to update the parameters with a call to step() of the optimizer. Replicating models in every forward calls greatly hurts the performance in [my implementation](https://github.com/anoidgit/transformer).

## Pitch

I tried to alleviate this through the following code:
```
class DataParallelModel(DataParallel):
    def __init__(self, module, device_ids=None, output_device=None, dim=0, host_control=False):
        super(DataParallelModel, self).__init__(module, device_ids, output_device, dim)

        if host_control:
	    self.update_replicates()
        else:
	    self.nets = None

    def forward(self, *inputs, **kwargs):
        if not self.device_ids:
            return self.module(*inputs, **kwargs)
        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
        if len(self.device_ids) == 1:
            return self.module(*inputs[0], **kwargs[0])
        if self.nets is None:
            replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
        else:
	    replicas = self.nets[:len(inputs)]
        outputs = self.parallel_apply(replicas, inputs, kwargs)
        return self.gather(outputs, self.output_device)

    def update_replicates(self):
        self.nets = self.replicate(self.module, self.device_ids)
```
Since I want to make the training more efficient like this:
```
model = DataParallelModel(designed_model, host_control=True)
crit = get_some_criterion()
optm = get_an_optimizer(model)
update_batches = 16
i = 1
for id, td in training_data():
    loss = crit(model(id), td)
    loss.backward()
    if i % update_batches == 0:
        optm.step()
        model.zero_grad()
        model.update_replicates()
    i += 1
```
But an assertation failure was thrown by autograd, which is beyond my ability.
I hope it can support another function like `accumulate_gradients` which accumulates the gradients from those replicated models at the same time, thus I can avoid useless communications between GPUs. In this case, the training code could be like:
```
model = DataParallelModel(designed_model, host_control=True)
crit = get_some_criterion()
optm = get_an_optimizer(model)
update_batches = 16
i = 1
for id, td in training_data():
    loss = crit(model(id), td)
    loss.backward()
    if i % update_batches == 0:
        model.accumulate_gradients()
        optm.step()
        model.zero_grad()
        model.update_replicates()
    i += 1
```
Accumulate the gradients of several small batches is very important in some cases where a large batch size is very important for the quality of the model while the GPU memory is not enough to support it, like the training of transformer(a Neural Machine Translation model).

**Comment by anoidgit** [comment]
I fix it with [this](https://github.com/anoidgit/transformer/blob/master/parallel.py):
```
#encoding: utf-8

import torch
import torch.cuda.comm as comm
from torch.nn import DataParallel

from threading import Lock, Thread

"""	Example::

		>>> net = DataParallelModel(model, device_ids=[0, 1, 2])
		>>> criterion = DataParallelCriterion(criterion, device_ids=[0, 1, 2])
		>>> y = net(x)
		>>> loss = criterion(y, target)
"""

# todo: control the replicating and parameters collecting in the future
class DataParallelModel(DataParallel):

	# host replicates should improve a little bit performance if there are additional calls to update_replicates and collect_gradients in the training scripts.
	def __init__(self, module, device_ids=None, output_device=None, dim=0, host_replicate=False):

		super(DataParallelModel, self).__init__(module, device_ids, output_device, dim)

		if host_replicate:
			self.update_replicates()
		else:
			self.nets = None

	def forward(self, *inputs, **kwargs):
		if not self.device_ids:
			return [self.module(*inputs, **kwargs)]
		inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
		if (len(self.device_ids) == 1) or (len(inputs) == 1):
			return [self.module(*inputs[0], **kwargs[0])]
		nbatch = len(inputs)
		devices = self.device_ids[:nbatch]
		if self.nets is None:
			replicas = self.replicate(self.module, devices)
		else:
			replicas = self.nets[:nbatch]
		outputs = parallel_apply(replicas, inputs, devices, kwargs)
		return outputs

	def update_replicates(self):
		self.nets = replicate_model(self.module, self.device_ids, True)

	def collect_gradients(self):
		grads = reduce_gradients([p.grad for net in self.nets for p in net.parameters()], len(self.device_ids), self.output_device)
		for mp, grad in zip(self.module.parameters(), grads):
			mp.grad = grad

class DataParallelCriterion(DataParallel):

	# if there is no parameter update in criterion, turn on replicate_once should improve a little bit performance.
	def __init__(self, module, device_ids=None, output_device=None, dim=0, replicate_once=False):

		super(DataParallelCriterion, self).__init__(module, device_ids, output_device, dim)

		if replicate_once:
			self.nets = replicate_model(self.module, self.device_ids, True)
		else:
			self.nets = None

	def forward(self, inputs, *targets, **kwargs):
		# input should be already scatterd
		# scattering the targets instead
		if not self.device_ids:
			return self.module(inputs, *targets, **kwargs)
		targets, kwargs = self.scatter(targets, kwargs, self.device_ids)
		if (len(self.device_ids) == 1) or (len(targets) == 1):
			return self.module(inputs[0], *targets[0], **kwargs[0])
		ngpu = len(inputs)
		devices = self.device_ids[:ngpu]
		if self.nets is None:
			replicas = self.replicate(self.module, devices)
		else:
			replicas = self.nets[:ngpu]
		outputs = criterion_parallel_apply(replicas, inputs, targets, devices, kwargs)

		return self.gather(outputs, self.output_device)

# update this functions with the update of replicate(https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py)

def reduce_gradients(grads, num_inputs, destination):
	return comm.reduce_add_coalesced([grads[i:i + num_inputs] for i in range(0, len(grads), num_inputs)], destination)

# update this functions with the update of replicate(https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/replicate.py)

def replicate_model(network, devices, no_gradient=False):

	def clear_gradient(para):
		if no_gradient:
			para.grad = None
		return para

	num_replicas = len(devices)

	params = [clear_gradient(para) for para in network.parameters()] if no_gradient else list(network.parameters())
	param_indices = {param: idx for idx, param in enumerate(params)}
	param_copies = tuple([t for tensors in comm.broadcast_coalesced(params, devices) for t in tensors])
	if len(params) > 0:
		param_copies = [param_copies[i:i + len(params)] for i in range(0, len(param_copies), len(params))]

	buffers = list(network.buffers())
	buffer_indices = {buf: idx for idx, buf in enumerate(buffers)}
	buffer_copies = comm.broadcast_coalesced(buffers, devices)

	modules = list(network.modules())
	module_copies = [[] for device in devices]
	module_indices = {}

	for i, module in enumerate(modules):
		module_indices[module] = i
		for j in range(num_replicas):
			replica = module.__new__(type(module))
			replica.__dict__ = module.__dict__.copy()
			replica._parameters = replica._parameters.copy()
			replica._buffers = replica._buffers.copy()
			replica._modules = replica._modules.copy()
			module_copies[j].append(replica)

	for i, module in enumerate(modules):
		for key, child in module._modules.items():
			if child is None:
				for j in range(num_replicas):
					replica = module_copies[j][i]
					replica._modules[key] = None
			else:
				module_idx = module_indices[child]
				for j in range(num_replicas):
					replica = module_copies[j][i]
					replica._modules[key] = module_copies[j][module_idx]
		for key, param in module._parameters.items():
			if param is None:
				for j in range(num_replicas):
					replica = module_copies[j][i]
					replica._parameters[key] = None
			else:
				param_idx = param_indices[param]
				requires_grad = param.requires_grad
				for j in range(num_replicas):
					replica = module_copies[j][i]
					replica._parameters[key] = param_copies[j][param_idx].requires_grad_(requires_grad)
		for key, buf in module._buffers.items():
			if buf is None:
				for j in range(num_replicas):
					replica = module_copies[j][i]
					replica._buffers[key] = None
			else:
				buffer_idx = buffer_indices[buf]
				for j in range(num_replicas):
					replica = module_copies[j][i]
					replica._buffers[key] = buffer_copies[j][buffer_idx]

	return [module_copies[j][0] for j in range(num_replicas)]

# update this functions with the update of parallel_apply(https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/parallel_apply.py)

def parallel_apply(modules, inputs, devices, kwargs_tup=None):
	if kwargs_tup is None:
		kwargs_tup = ({},) * len(modules)

	lock = Lock()
	results = {}

	def _worker(i, module, input, kwargs, device=None):

		with torch.cuda.device(device):
			# this also avoids accidental slicing of `input` if it is a Tensor
			if not isinstance(input, (list, tuple)):
				input = (input,)
			output = module(*input, **kwargs)
		with lock:
			results[i] = output

	threads = [Thread(target=_worker, args=(i, module, input, kwargs, device)) for i, (module, input, kwargs, device) in enumerate(zip(modules, inputs, kwargs_tup, devices))]

	for thread in threads:
		thread.start()
	for thread in threads:
		thread.join()

	outputs = []
	for i in range(len(inputs)):
		output = results[i]
		outputs.append(output)
	return outputs

def criterion_parallel_apply(modules, inputs, targets, devices, kwargs_tup=None):

	if kwargs_tup is None:
		kwargs_tup = ({},) * len(modules)

	lock = Lock()
	results = {}

	def _worker(i, module, input, target, kwargs, device):
		with torch.cuda.device(device):
			# this also avoids accidental slicing of `input` if it is a Tensor
			if not isinstance(input, (list, tuple)):
				input = (input,)
			if not isinstance(target, (list, tuple)):
				target = (target,)
			output = module(*(input + target), **kwargs)
		with lock:
			results[i] = output

	threads = [Thread(target=_worker, args=(i, module, input, target, kwargs, device)) for i, (module, input, target, kwargs, device) in enumerate(zip(modules, inputs, targets, kwargs_tup, devices))]

	for thread in threads:
		thread.start()
	for thread in threads:
		thread.join()

	outputs = []
	for i in range(len(inputs)):
		output = results[i]
		outputs.append(output)
	return outputs
```
